### PR TITLE
fix: use lenient matching for coverage merge when end.column differs

### LINKED ISSUE #719: Combining Code Coverage (Karma & Cypress)
We are noticing some odd things when trying to combine the coverage. 
Looking at the JSON report for the statementMap
Both commonly have a null for the end column but in some instances karma will report a null and cypress (with istanbul instrumentation) will report a specific end column number. When these are combined they are treated as 2 different statements for the coverage stats
`{
			"start": {
				"line": 16,
				"column": 4
			},
			"end": {
				"line": 16,
				"column": null
			}
		}`
vs
`{
			"start": {
				"line": 16,
				"column": 4
			},
			"end": {
				"line": 16,
				"column": 46
			}
		}`

What does a null column end mean? Could the merge handle this more gracefully?

In this case 16 was wrapped in a if statement that the cypress coverage picked up on for the statementMap and also the branchMap but the karma coverage omited from both.
`{
			"start": {
				"line": 15,
				"column": 2
			},
			"end": {
				"line": 17,
				"column": null
			}
		}`

We also saw the same thing happen outside the if statement
`{
			"start": {
				"line": 18,
				"column": 2
			},
			"end": {
				"line": 18,
				"column": 14
			}
		}`
vs
`{
			"start": {
				"line": 18,
				"column": 2
			},
			"end": {
				"line": 18,
				"column": null
			}
		}`

And also the same null ending column issue in the fnMap
`{
			"name": "getError",
			"decl": {
				"start": {
					"line": 14,
					"column": 16
				},
				"end": {
					"line": 14,
					"column": 24
				}
			},
			"loc": {
				"start": {
					"line": 14,
					"column": 45
				},
				"end": {
					"line": 19,
					"column": null
				}
			}
		}`
vs
`{
			"name": "getError",
			"decl": {
				"start": {
					"line": 14,
					"column": 16
				},
				"end": {
					"line": 14,
					"column": 24
				}
			},
			"loc": {
				"start": {
					"line": 14,
					"column": 45
				},
				"end": {
					"line": 19,
					"column": 1
				}
			}`

### PR DESCRIPTION
**Pull request by stevez**
When merging coverage from different transpilers (e.g., Vitest jsdom vs browser tests), the end.column values can differ for the same logical code location. This caused duplicate entries instead of proper merging.

This fix adds lenient matching that ignores end.column when exact match fails. The matching now uses a two-pass approach:
1. Try exact match (start.line|start.column|end.line|end.column)
2. Fall back to lenient match (start.line|start.column|end.line)

This affects statements, functions, and branches via the mergeProp function.

Fixes #719

### CODE REVIEW

File: packages/istanbul-lib-coverage/lib/file-coverage.js [hunk 1]
```diff
@@ -45,6 +45,11 @@ function assertValidObject(obj) {
 const keyFromLoc = ({ start, end }) =>
     `${start.line}|${start.column}|${end.line}|${end.column}`;
 
+// Lenient key that ignores end.column - used for fuzzy matching when exact match fails
+// This handles cases where different transpilers produce different end.column values
+const keyFromLocLenient = ({ start, end }) =>
+    `${start.line}|${start.column}|${end.line}`;
```
Comment (AriPerkkio) [inline packages/istanbul-lib-coverage/lib/file-coverage.js]: This line crashes on Vitest's internal test cases. `mergeProp` here

https://github.com/istanbuljs/istanbuljs/blob/28ffdbc314596bdcb3007e85d30a62372602b262/packages/istanbul-lib-coverage/lib/file-coverage.js#L305-L310

... is called with:

```js
mergeProp(
  // this.s:
  { '0': 1, '1': 0, '2': 0, '3': 0 },
  // this.statementMap:
  {
    '0': { start: { line: 2, column: 2 }, end: { line: 2, column: null } },
    '1': { start: { line: 6, column: 2 }, end: { line: 6, column: null } },
    '2': { start: { line: 10, column: 2 }, end: { line: 10, column: null } },
    '3': { start: { line: 14, column: 2 }, end: { line: 14, column: null } }
  },
  // other.s:
  { '0': 0, '1': 0, '2': 1, '3': 0 },
  // other.statementMap:
  {
    '0': { start: { line: 2, column: 2 }, end: { line: 2, column: null } },
    '1': { start: { line: 6, column: 2 }, end: { line: 6, column: null } },
    '2': { start: { line: 10, column: 2 }, end: { line: 10, column: null } },
    '3': { start: { line: 14, column: 2 }, end: { line: 14, column: null } }
  }
);
```

This crashes with

```
TypeError: Cannot destructure property 'start' of 'object null' as it is null.
 ❯ keyFromLocLenient ../../packages/coverage-v8/dist/provider.js:237:31
 ❯ itemKeyLenient ../../packages/coverage-v8/dist/provider.js:330:17
 ❯ buildItemsIndex ../../packages/coverage-v8/dist/provider.js:341:33
 ❯ mergeProp ../../packages/coverage-v8/dist/provider.js:349:38
 ❯ FileCoverage.merge ../../packages/coverage-v8/dist/provider.js:538:21
 ❯ CoverageMap.addFileCoverage ../../packages/coverage-v8/dist/provider.js:804:30
 ❯ ../../packages/coverage-v8/dist/provider.js:747:19
 ❯ CoverageMap.merge ../../packages/coverage-v8/dist/provider.js:746:36
 ❯ V8CoverageProvider.mergeReports ../../packages/vitest/dist/chunks/coverage.BU-ylMbC.js:3177:52
```

Crash happens on https://github.com/vitest-dev/vitest/blob/main/test/coverage-test/test/merge-reports.test.ts.

**Review by AriPerkkio — CHANGES_REQUESTED** [review]

File: packages/istanbul-lib-coverage/lib/file-coverage.js (hunk 1 above)
Comment (stevez) [reply packages/istanbul-lib-coverage/lib/file-coverage.js]: @AriPerkkio , thanks for the feedback, the PR is updated and the crash issue is fixed

Comment (AriPerkkio) [reply packages/istanbul-lib-coverage/lib/file-coverage.js]: Those changes fixed the crash.

Comment (stevez) [reply packages/istanbul-lib-coverage/lib/file-coverage.js]: @AriPerkkio could you approve this PR?

Comment (AriPerkkio) [reply packages/istanbul-lib-coverage/lib/file-coverage.js]: I need to review the code first.

Comment (stevez) [reply packages/istanbul-lib-coverage/lib/file-coverage.js]: let me know if you have any questions regarding the code

File: packages/istanbul-lib-coverage/lib/file-coverage.js [hunk 2]
```diff
@@ -131,36 +136,76 @@ const addNearestContainerHits = (item, itemHits, map, mapHits) => {
...
-        const item = bMap[key];
-        bItems[itemKey(item)] = [itemHits, item];
-    }
+    // Derive lenient key function from the provided itemKey
+    // Handle both regular items (statements, functions) and branch items (which use locations[0])
+    // Returns null if loc is invalid (e.g., end.column is null)
+    const itemKeyLenient = item => {
+        const loc = item.locations ? getLoc(item.locations[0]) : getLoc(item);
+        return loc ? keyFromLocLenient(loc) : null;
+    };
```
Comment (AriPerkkio) [inline packages/istanbul-lib-coverage/lib/file-coverage.js]: Could we just re-use the result from `itemKey` and ignore the end mapping?

```js
const REMOVE_END_MAPPING_PATTERN = /^([^|]+\|[^|]+\|[^|]+)\|[^|]*$/;

...

const mergeProp = (aHits, aMap, bHits, bMap, itemKey = keyFromLoc) => {
    // Derive lenient key ...
    const itemKeyLenient = item => itemKey(item).replace(REMOVE_END_MAPPING_PATTERN, '$1');
```

File: packages/istanbul-lib-coverage/test/file.test.js [hunk 3]
```diff
@@ -1124,6 +1124,241 @@ describe('addNearestContainerHits unit coverage', () => {
...
+        assert.deepEqual(summary.branches, {
+            total: 2,
+            covered: 2,
+            skipped: 0,
+            pct: 100
+        });
+    });
+
+    it('only stores first A item when multiple A items share same lenient key', () => {
+        // Test coverage for lines 155-157: if (!aItemsLenient[lenientKey])
```
Comment (AriPerkkio) [inline packages/istanbul-lib-coverage/test/file.test.js]: Comments that refence line numbers or source code become outdated easily. Let's just remove these:


```suggestion
```

File: packages/istanbul-lib-coverage/test/file.test.js [hunk 4]
```diff
@@ -1124,6 +1124,241 @@ describe('addNearestContainerHits unit coverage', () => {
...
+                start: { line: sl, column: sc },
+                end: { line: el, column: ec }
+            };
+        };
+
+        const c1 = new FileCoverage({
+            path: '/path/to/file',
+            statementMap: {
+                0: loc(1, 0, 1, 50), // lenient key: 1|0|1
+                1: loc(1, 0, 1, 60) // lenient key: 1|0|1 (same!) - triggers line 155 false branch
```
Comment (AriPerkkio) [inline packages/istanbul-lib-coverage/test/file.test.js]:
```suggestion
                1: loc(1, 0, 1, 60) // lenient key: 1|0|1 (same!)
```

File: packages/istanbul-lib-coverage/test/file.test.js [hunk 5]
```diff
@@ -1124,6 +1124,241 @@ describe('addNearestContainerHits unit coverage', () => {
...
+        assert.equal(c1.s[1], 3);
+        const summary = c1.toSummary();
+        assert.equal(summary.statements.total, 2);
+    });
+
+    it('skips extra B items that share lenient key with A items', () => {
+        // Test coverage for line 210: if (aItemsLenient[lenientKey]) continue;
+        // When B has multiple items with same lenient key as an A item,
+        // only the first B item gets lenient-matched in the first loop.
+        // The second B item should be skipped in the second loop via line 210.
```
Comment (AriPerkkio) [inline packages/istanbul-lib-coverage/test/file.test.js]:
```suggestion
```

File: packages/istanbul-lib-coverage/test/file.test.js [hunk 6]
```diff
@@ -1124,6 +1124,241 @@ describe('addNearestContainerHits unit coverage', () => {
...
+            s: { 0: 5 },
+            f: {},
+            b: {}
+        });
+
+        const c2 = new FileCoverage({
+            path: '/path/to/file',
+            statementMap: {
+                0: loc(1, 0, 1, 99), // lenient key: 1|0|1 (same as c1)
+                1: loc(1, 0, 1, 88) // lenient key: 1|0|1 (same as c1) - this triggers line 210
```
Comment (AriPerkkio) [inline packages/istanbul-lib-coverage/test/file.test.js]:
```suggestion
                1: loc(1, 0, 1, 88) // lenient key: 1|0|1 (same as c1)
```

File: packages/istanbul-lib-coverage/test/file.test.js [hunk 7]
```diff
@@ -1124,6 +1124,241 @@ describe('addNearestContainerHits unit coverage', () => {
...
+            branchMap: {},
+            s: { 0: 3, 1: 7 },
+            f: {},
+            b: {}
+        });
+
+        c1.merge(c2);
+
+        // A's item should be merged with B's first matching item (index 0)
+        // B's second item (index 1) shares same lenient key, should be skipped via line 210
```
Comment (AriPerkkio) [inline packages/istanbul-lib-coverage/test/file.test.js]:
```suggestion
```

**Review by AriPerkkio — CHANGES_REQUESTED** [review]
Overall looks good even though it's quite complex. The test cases seem to validate edge cases too. I tried to simplify the implementation but started seeing failing test cases quickly. 👍

File: packages/istanbul-lib-coverage/test/file.test.js (hunk 7 above)
Comment (stevez) [reply packages/istanbul-lib-coverage/test/file.test.js]: @AriPerkkio, just update the PR based on your suggestion: 1)derive lenient key from exact key using regex; 2) removed the comments from the file.test.js

**Review by AriPerkkio — APPROVED** [review]
Looks good to me 👍

**Review by AriPerkkio — CHANGES_REQUESTED** [review]
Marking back to `Request changes` as merging functions still produces duplicates:
- https://github.com/vitest-dev/vitest/pull/9480#discussion_r2718326878
